I'd say let's address this first so we get rid of the warning:
```
.../llvm-project/clang/lib/Index/IndexDataStoreUtils.cpp:286:11: warning: enumeration values 'TemplateTypeParm', 'TemplateTemplateParm', and 'NonTypeTemplateParm' not handled in switch [-Wswitch]
```

Couple thoughts re: splitting `SymbolKind`/`SymbolSubKind` into more generic/more language specific kinds.
- The enums are mostly used the way you suggest but not completely.
- used both in libclang and clangd
- There might be less intrusive ways (less maintenance friction) how to do this?

--- a/clang/lib/Index/IndexDataStoreUtils.cpp
+++ b/clang/lib/Index/IndexDataStoreUtils.cpp
@@ -285,6 +285,9 @@ SymbolRoleSet index::getSymbolRoles(uint64_t Roles) {
 indexstore_symbol_kind_t index::getIndexStoreKind(SymbolKind K) {
   switch (K) {
   case SymbolKind::Unknown:
+  case SymbolKind::TemplateTypeParm:
+  case SymbolKind::TemplateTemplateParm:
+  case SymbolKind::NonTypeTemplateParm:
     return INDEXSTORE_SYMBOL_KIND_UNKNOWN;
   case SymbolKind::Module:
     return INDEXSTORE_SYMBOL_KIND_MODULE;
@@ -335,9 +338,6 @@ indexstore_symbol_kind_t index::getIndexStoreKind(SymbolKind K) {
   case SymbolKind::ConversionFunction:
     return INDEXSTORE_SYMBOL_KIND_CONVERSIONFUNCTION;
   case SymbolKind::Parameter:
-  case SymbolKind::TemplateTypeParm:
-  case SymbolKind::TemplateTemplateParm:
-  case SymbolKind::NonTypeTemplateParm:
     return INDEXSTORE_SYMBOL_KIND_PARAMETER;
   case SymbolKind::Using:
     return INDEXSTORE_SYMBOL_KIND_USING;

--- a/clang/lib/Index/IndexDataStoreUtils.cpp
+++ b/clang/lib/Index/IndexDataStoreUtils.cpp
@@ -285,6 +285,19 @@ SymbolRoleSet index::getSymbolRoles(uint64_t Roles) {
 indexstore_symbol_kind_t index::getIndexStoreKind(SymbolKind K) {
   switch (K) {
   case SymbolKind::Unknown:
+  // FIXME: Tweak SymbolKind/SymbolSubKind to separate
+  // generic-ish/language-specific-ish concepts.
+  //
+  // We use SymbolKind/SymbolSubKind
+  // downstream for Swift and this would help us avoid intrusions here or
+  // upstream.
+  //
+  // Separation of concepts for C++ and Objective-C could also be
+  // improved. Once that is done on SymbolKind/SymbolSubKind level we should
+  // update indexstore_symbol_kind_t too.
+  //
+  // Initial discussion here:
+  // https://github.com/apple/llvm-project/pull/1099
   case SymbolKind::TemplateTypeParm:
   case SymbolKind::TemplateTemplateParm:
   case SymbolKind::NonTypeTemplateParm:

